rfcs/0013-changes-lifecyle.md

```diff
@@ -0,0 +1,348 @@
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+This section contains the suggested guide on how to introduce changes to different parts of the compiler and associated tools.
+
+## `sway-features` crate
+
+A new crate named `sway-features` will be created and will contain **ALL** features and their metadata. The best suggestion is a macro to define features where enums, documentation, etc., will be generated.
+
+This macro also needs to generate code for the errors and warnings related to each error.
+
+Features will be parsed in this order:
+
+```rust
+let mut features = ExperimentalFeatures::default();
+features.parse_from_toml();
+features.parse_from_cli();
+features.parse_from_environment_variables();
+```
+
+This means that environment variables overwrite CLI arguments, which overwrite TOML configuration.
+
+If a feature requires or allows new configurations, these configurations should be optional, and as soon as confirmed the feature is enabled, it should verify if the required configuration is available.
+
+```
+> forc ... --experimental a_feature --a-feature-option 1
+```
+
+## Lexer and Parser
+
+To allow as user-friendly error messages as possible, in all possible cases, we want both the lexer and the parser to parse the new syntax even with the feature off.
+
+After that, the lexer and the parser will mark that an experimental feature was lexed/parsed; and a check will guarantee that no disabled experimental was parsed.
+
+The error message will have a message explaining that the feature is experimental and a GitHub link for more details on the stabilization lifecycle.
+
+```rust
+// always parse new syntax
+let new_syntax = parse_new_syntax();
+if ctx.experimental.is_disabled(Features::NEW_FEATURE) {
+   handler.emit_error(...);
+}
+```
+
+## Formatting
+
+Formatting should always format new syntax, even when the flag is off. To avoid breaking other parts of the code, or even worse, removing code.
+
+## LSP
+
+LSP can take advantage of specific error messages, and suggest users to enable the corresponding feature.
+
+## CST, AST, and Typed Tree
+
+All trees must always support new features, which means that new nodes will always exist. Their specific behavior, desugaring, etc., will be gated by the experimental feature.
+
+More specifically, this means that variants should not be behind compiler flags.
+
+## Tests
+
+To allow these experimental features to be tested, it is necessary to specify which configurations each test must run.
+Today, by default, a test runs with the "default compiler configuration", but it is necessary to allow tests to
+run with any combination of compiler configuration.
+
+To allow complete tests of any compiler configuration, any configuration that today resides inside `test.toml`
+needs to be configured for each configuration.
+
+An example is the `encoding v1` which has different inputs and outputs from `encoding v0` when testing a script
```

Comment:
This section reads more like a possible proposal, whereas RFCs should describe the state of the system as if it's already in place. That way when it is merged and referenced later on, the expectations and specifications are clear to the user.